src/analysis/seminorm.lean

```diff
@@ -34,17 +38,171 @@ For a module over a normed ring:
+noncomputable instance : has_sup (add_monoid_seminorm E) :=
+{ sup := λ p q,
+  { to_fun  := p ⊔ q,
+    nonneg' := λ x, begin
+      simp only [pi.sup_apply, le_sup_iff],
+      exact or.intro_left _ (p.nonneg _),
+    end,
+    map_zero' := begin
+      simp only [pi.sup_apply],
+      rw [← p.map_zero, sup_eq_left, p.map_zero, q.map_zero],
+    end,
+    add_le' := λ x y, sup_le
+      ((p.add_le x y).trans $ add_le_add le_sup_left le_sup_left)
+      ((q.add_le x y).trans $ add_le_add le_sup_right le_sup_right) }}
+
+@[simp] lemma coe_sup (p q : add_monoid_seminorm E) : ⇑(p ⊔ q) = p ⊔ q := rfl
+lemma sup_apply (p q : add_monoid_seminorm E) (x : E) : (p ⊔ q) x = p x ⊔ q x := rfl
+
+lemma smul_sup [has_scalar R ℝ] [has_scalar R ℝ≥0] [is_scalar_tower R ℝ≥0 ℝ]
+  (r : R) (p q : add_monoid_seminorm E) :
+  r • (p ⊔ q) = r • p ⊔ r • q :=
+have real.smul_max : ∀ x y : ℝ, r • max x y = max (r • x) (r • y),
+from λ x y, by simpa only [←smul_eq_mul, ←nnreal.smul_def, smul_one_smul ℝ≥0 r (_ : ℝ)]
+                     using mul_max_of_nonneg x y (r • 1 : ℝ≥0).prop,
+ext $ λ x, real.smul_max _ _
+
+instance : partial_order (add_monoid_seminorm E) :=
+  partial_order.lift _ fun_like.coe_injective
+
+lemma le_def (p q : add_monoid_seminorm E) : p ≤ q ↔ (p : E → ℝ) ≤ q := iff.rfl
+lemma lt_def (p q : add_monoid_seminorm E) : p < q ↔ (p : E → ℝ) < q := iff.rfl
+
+noncomputable instance : semilattice_sup (add_monoid_seminorm E) :=
+function.injective.semilattice_sup _ fun_like.coe_injective coe_sup
+
+end add_monoid_seminorm
+
 /-- A seminorm on a module over a normed ring is a function to the reals that is positive
 semidefinite, positive homogeneous, and subadditive. -/
-structure seminorm (𝕜 : Type*) (E : Type*) [semi_normed_ring 𝕜] [add_monoid E] [has_scalar 𝕜 E] :=
-(to_fun    : E → ℝ)
-(smul'     : ∀ (a : 𝕜) (x : E), to_fun (a • x) = ∥a∥ * to_fun x)
-(triangle' : ∀ x y : E, to_fun (x + y) ≤ to_fun x + to_fun y)
+structure seminorm (𝕜 : Type*) (E : Type*) [semi_normed_ring 𝕜] [add_monoid E] [has_scalar 𝕜 E]
+  extends add_monoid_seminorm E :=
+(smul' : ∀ (a : 𝕜) (x : E), to_fun (a • x) = ∥a∥ * to_fun x)
+
+attribute [nolint doc_blame] seminorm.to_add_monoid_seminorm
+
+private lemma map_zero.of_smul {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_monoid E]
+  [smul_with_zero 𝕜 E] {f : E → ℝ} (smul : ∀ (a : 𝕜) (x : E), f (a • x) = ∥a∥ * f x) : f 0 = 0 :=
+calc f 0 = f ((0 : 𝕜) • 0) : by rw zero_smul
+     ... = 0 : by rw [smul, norm_zero, zero_mul]
+
+private lemma neg.of_smul {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_comm_group E]
+  [module 𝕜 E] {f : E → ℝ} (smul : ∀ (a : 𝕜) (x : E), f (a • x) = ∥a∥ * f x)
+  (x : E) : f (-x) = f x :=
+by rw [←neg_one_smul 𝕜, smul, norm_neg, ← smul, one_smul]
+
+private lemma nonneg.of {𝕜 : Type*} {E : Type*} [semi_normed_ring 𝕜] [add_comm_group E] [module 𝕜 E]
+  [norm_one_class 𝕜] {f : E → ℝ} (add_le : ∀ (x y : E), f (x + y) ≤ f x + f y)
```

Comment:
I think my earlier comment got lost; this doesn't need `norm_one_class`, and indeed the proof this replaces in master already doesn't use it

Reply:
It seems I forgot to remove this from two lemmas. It should be done now, thanks!